File: front/src/types/datasets.ts
// ============================================================================
// Dataset Types
// ============================================================================

// ----------------------------------------------------------------------------
// Domain Primitives
// ----------------------------------------------------------------------------

type UUID = string;
type ISO8601Date = string;
type URL = string;

// ----------------------------------------------------------------------------
// Dataset Domain Types
// ----------------------------------------------------------------------------

type Publisher = string | null;
type DatasetTitle = string | null;

type QualityIndicators = {
    has_description: boolean;
    is_slug_valid: boolean;  // No underscores or special chars
    evaluation_results: any | null;  // LLM evaluation results (JSON)
}

// ----------------------------------------------------------------------------
// List View (Table Display)
// ----------------------------------------------------------------------------

export type DatasetSummary = {
    id: UUID;
    platformId: UUID;
    platformName?: string;
    platformType?: string;
    publisher: Publisher;
    title: DatasetTitle;
    created: ISO8601Date;
    modified: ISO8601Date;
    downloadsCount?: number;
    apiCallsCount?: number;
    viewsCount?: number;
    reusesCount?: number;
    followersCount?: number;
    popularityScore?: number;
    versionsCount?: number;
    page: URL;
    restricted: boolean | null;
    published: boolean | null;
    lastSyncStatus: string;
    hasDescription: boolean;
    isDeleted: boolean | null;
    quality?: QualityIndicators;
};

// ----------------------------------------------------------------------------
// Version History
// ----------------------------------------------------------------------------

/**
 * Immutable snapshot of dataset state at a point in time.
 * Backend pre-calculates diffs for performance.
 */
export type SnapshotVersion = {
    id: UUID;
    timestamp: ISO8601Date;
    downloadsCount: number | null;
    apiCallsCount: number | null;
    viewsCount?: number | null;
    reusesCount?: number | null;
    followersCount?: number | null;
    popularityScore?: number | null;
    page: URL;
    title?: DatasetTitle;
    data?: unknown;  // Full snapshot (only when include_data=true)
    diff?: any;  // Pre-calculated by backend
};

// ----------------------------------------------------------------------------
// Detail View (Modal Display)
// ----------------------------------------------------------------------------

export type DatasetDetail = {
    id: UUID;
    platformId: UUID;
    publisher: Publisher;
    title: DatasetTitle;
    buid: string;  // Business UID from source platform
    slug: string;
    page: URL;
    created: ISO8601Date;
    modified: ISO8601Date;
    published: boolean | null;
    restricted: boolean | null;
    hasDescription: boolean;
    isDeleted: boolean | null;
    quality?: QualityIndicators;
    currentSnapshot: SnapshotVersion | null;
    snapshots?: SnapshotVersion[] | null;  // Only when include_snapshots=true
};

// ----------------------------------------------------------------------------
// API Contracts
// ----------------------------------------------------------------------------

export type PaginatedResponse<T> = {
    items: T[];
    total: number;
    page: number;  // 1-indexed
    pageSize: number;
};

export type DatasetListQuery = {
    // Filters
    platformId?: UUID;
    publisher?: string;  // Partial match
    q?: string;  // Searches slug only
    isDeleted?: boolean;

    // Date ranges
    createdFrom?: ISO8601Date;
    createdTo?: ISO8601Date;
    modifiedFrom?: ISO8601Date;
    modifiedTo?: ISO8601Date;

    // Sorting
    sortBy?: 'created' | 'modified' | 'publisher' | 'downloads_count' | 'api_calls_count' | 'title' | 'versions_count' | 'popularity_score' | 'views_count' | 'reuses_count';
    order?: 'asc' | 'desc';

    // Pagination
    page?: number;
    pageSize?: number;
};

// ----------------------------------------------------------------------------
// Publishers
// ----------------------------------------------------------------------------

/**
 * List of all publishers in the system.
 * Used for autocomplete and filtering.
 */
export type PublishersRef = string[];

// ----------------------------------------------------------------------------
// Re-exports
// ----------------------------------------------------------------------------

export type { PlatformRef, PlatformSync } from './platforms';

File: src/interfaces/api/schemas/datasets.py
"""
Dataset API Schemas
Self-documenting schemas using type annotations and minimal descriptions.
"""

import datetime
from typing import Annotated
from uuid import UUID

from pydantic import BaseModel, Field

# ============================================================================
# Domain Types
# ============================================================================

DatasetId = UUID
BusinessUID = str  # Unique ID from source platform (e.g., "ods:eau-2023")
DatasetSlug = str  # URL-friendly identifier
SourceURL = str
Publisher = str | None
SyncStatus = str  # 'pending' | 'success' | 'failed'

# Validated types
PositiveInt = Annotated[int, Field(ge=0)]
Score = Annotated[float, Field(ge=0.0, le=100.0)]

# ============================================================================
# List View
# ============================================================================


class DatasetAPI(BaseModel):
    """Dataset summary for list/table display."""

    # Identity
    id: DatasetId | None = None
    buid: BusinessUID
    slug: DatasetSlug
    title: str | None = None

    # Source
    page: SourceURL
    publisher: Publisher = None

    # Timestamps
    timestamp: datetime.datetime | None = None  # Last snapshot
    created: datetime.datetime
    modified: datetime.datetime

    # Flags
    published: bool
    restricted: bool
    deleted: bool = False  # Soft delete flag

    # Metrics (populated when include_counts=true)
    downloads_count: PositiveInt | None = None
    api_calls_count: PositiveInt | None = None
    views_count: PositiveInt | None = None
    reuses_count: PositiveInt | None = None
    followers_count: PositiveInt | None = None
    popularity_score: Score | None = None
    versions_count: PositiveInt = 0

    # Sync status
    last_sync: datetime.datetime | None = None
    last_sync_status: SyncStatus


# ============================================================================
# Dataset Creation
# ============================================================================


class DatasetCreateResponse(BaseModel):
    """Response after creating a new dataset via API."""

    id: DatasetId
    name: str
    timestamp: str
    buid: BusinessUID
    slug: DatasetSlug
    organization_id: str
    type: str  # Platform type
    url: SourceURL
    key: str  # API key used


# ============================================================================
# Version History
# ============================================================================


class SnapshotVersionAPI(BaseModel):
    """
    Immutable snapshot of dataset state at a point in time.
    Diffs are pre-calculated by backend for performance.
    """

    id: UUID
    timestamp: datetime.datetime
    title: str | None = None

    # Metrics at this point in time
    downloads_count: PositiveInt | None = None
    api_calls_count: PositiveInt | None = None
    views_count: PositiveInt | None = None
    reuses_count: PositiveInt | None = None
    followers_count: PositiveInt | None = None
    popularity_score: Score | None = None

    # Backend-calculated diff from previous version
    diff: dict | None = None

    # Full snapshot data (only when include_data=true)
    data: dict | None = None


# ============================================================================
# API Responses
# ============================================================================


# ============================================================================
# API Responses
# ============================================================================


class DatasetResponse(BaseModel):
    """Paginated list of datasets."""

    datasets: list[DatasetAPI]
    total_datasets: int


class DatasetDetailAPI(DatasetAPI):
    """Comprehensive dataset view with detailed quality and snapshots."""

    quality: dict | None = None
    current_snapshot: SnapshotVersionAPI | None = None
    snapshots: list[SnapshotVersionAPI] | None = None


class DatasetVersionsResponse(BaseModel):
    """Paginated list of dataset versions."""

    items: list[SnapshotVersionAPI]
    total: int
    page: int  # 1-indexed
    page_size: int
